tools/modlist.d

```diff
@@ -4,147 +4,143 @@
+                    writeln("$(MENU_W_SUBMENU_END)");
+                }
+                else
+                    writef("\n%s)", indent(lastPackage.length));
+            }
+            while (lastPackage.length < pkg.length)
+            {
+                lastPackage = pkg[0 .. lastPackage.length + 1];
+                if (lastPackage.length == 1) // top-level
+                {
+                    writeln();
+                    writefln("$(MENU_W_SUBMENU %s)", lastPackage[0]);
+                    writef("$(ITEMIZE");
+                }
+                else
+                    writef("%s\n%s$(PACKAGE $(PACKAGE_NAME %s),",
+                        firstChild ? "" : ",", indent(lastPackage.length - 1), lastPackage[$-1]);
+                firstChild = true;
+            }
+        }
+
+        void put(string[] mod, bool hasChildren)
+        {
+            needPackage(mod[0 .. $-1]);
+            writef("%s\n", firstChild ? "" : ",");
+            firstChild = false;
+            if (hasChildren)
+            {
+                writef("%s$(PACKAGE $(MODULE%s %-(%s, %))",
+                    indent(mod.length - 1), mod.length, mod);
+                lastPackage = mod;
+            }
+            else if (mod.length == 1) // top-level
+                writefln("$(MENU %s.html, $(TT %1$s))", mod[0]);
+            else
+                writef("%s$(MODULE%s %-(%s, %))",
+                    indent(mod.length - 1), mod.length, mod);
+        }
+
+        private static string indent(size_t len)
+        {
+            static immutable spaces = "                    ";
+            return spaces[0 .. 2 * len];
+        }
+
+        @disable this(this);
+        ~this() { needPackage(null); }
     }
 
-    tree.sort();
+    auto includedSet = packages.map!(pkg => tuple(pkg.split(".").assumeUnique, true)).assocArray;
+    auto excludedSet = excludes.map!(pkg => tuple(pkg.split(".").assumeUnique, true)).assocArray;
+    auto internalSet = internal.map!(pkg => tuple(pkg.split(".").assumeUnique, true)).assocArray;
 
     writeln("MODULE_MENU=");
-    foreach (part; packages)
+
+    foreach (dumpingInternal; [false, true])
     {
-        // check whether it's a package or file
-        auto subTree = tree[part];
-        if (subTree.leaves.length)
-            subTree.dumpRoot;
-        else
-            writefln("$(MENU %s.html, $(TT %1$s))", part);
+        if (dumpingInternal)
+            writeln("$(MENU_INTERNAL_SEPARATOR)");
+
+        foreach (pkg; packages)
+        {
+            TreeWriter writer;
+
+            void dump(ref Tree tree, string[] mod, bool inInternal, bool inIncluded)
+            {
+                inInternal = inInternal || mod in internalSet;
+                inIncluded = mod in includedSet ? true : mod in excludedSet ? false : inIncluded;
+                if (tree.exists && inInternal == dumpingInternal && inIncluded)
+                    if (mod.length > 1 || !tree.children) // std/package.d is currently unreachable!
```

Comment:
There's nothing one can click right now to see the `std/package.d` docs, because clicking on `std` expands/collapses the `std` tree.